source/draw/gpu/GpuSurface.ooc

```diff
@@ -16,14 +16,110 @@
+	_model: FloatTransform3D
+	_view: FloatTransform3D
+	_projection: FloatTransform3D
+	_toLocal: FloatTransform3D
+	transform: FloatTransform3D {
+		get { this _toLocal * this _view * this _toLocal }
+		set(value) { this _view = this _toLocal * value * this _toLocal }
+	}
+	_focalLength: Float
+	focalLength: Float {
+		get { this _focalLength }
+		set(value) {
+			this _focalLength = value
+			if (this _focalLength > 0.0f) {
+				a := 2.0f * this _focalLength / this size width
+				f := -(this _coordinateTransform e as Float) * 2.0f * this _focalLength / this size height
+				k := (this farPlane + this nearPlane) / (this farPlane - this nearPlane)
+				o := 2.0f * this farPlane * this nearPlane / (this farPlane - this nearPlane)
+				this _projection = FloatTransform3D new(a, 0.0f, 0.0f, 0.0f, 0.0f, f, 0.0f, 0.0f, 0.0f, 0.0f, k, -1.0f, 0.0f, 0.0f, o, 0.0f)
+			}
+			else
+				this _projection = FloatTransform3D createScaling(2.0f / this size width, -(this _coordinateTransform e as Float) * 2.0f / this size height, 1.0f)
+			this _model = this _createModelTransform(IntBox2D new(this size))
+		}
+	}
+	nearPlane: Float { get set }
+	farPlane: Float { get set }
+	_defaultMap: GpuMap
+	_coordinateTransform := IntTransform2D identity
+	init: func (=_size, =_context, =_defaultMap, =_coordinateTransform) {
+		this _toLocal = FloatTransform3D createScaling(1.0f, -1.0f, -1.0f)
+		this clearColor = ColorBgra new(0, 0, 0, 0)
+		this viewport = IntBox2D new(this size)
+		this focalLength = 0.0f
+		this nearPlane = 1.0f
+		this farPlane = 10000.0f
+		this _view = FloatTransform3D identity
+		this blend = false
+	}
+	_createModelTransform: func (box: IntBox2D) -> FloatTransform3D {
+		toReference := FloatTransform3D createTranslation((box size width - this size width) / 2, (this size height - box size height) / 2, 0.0f)
+		translation := this _toLocal * FloatTransform3D createTranslation(box leftTop x, box leftTop y, this focalLength) * this _toLocal
+ 		translation * toReference * FloatTransform3D createScaling(box size width / 2.0f, box size height / 2.0f, 1.0f)
+	}
+	_createTextureTransform: static func (imageSize: IntSize2D, box: IntBox2D) -> FloatTransform3D {
+		scaling := FloatTransform3D createScaling(box size width as Float / imageSize width, box size height as Float / imageSize height, 1.0f)
+		translation := FloatTransform3D createTranslation(box leftTop x as Float / imageSize width, box leftTop y as Float / imageSize height, 0.0f)
+		translation * scaling
+	}
+	_getDefaultMap: virtual func (image: Image) -> GpuMap { this _defaultMap }
+	_bind: virtual func
+	_unbind: virtual func
 	clear: abstract func
-	recycle: abstract func
+	draw: func ~general (action: Func) {
+		this _bind()
+		this _context setViewport(this viewport)
+		this _context enableBlend(this blend)
+		action()
+		this _unbind()
+	}
+	draw: func ~WithoutBind (destination: IntBox2D, map: GpuMap) {
```

Comment:
This function causes a lot of leaks, according to valgrind. My guess is the `func` needs to be freed after use.